### PR TITLE
Ground work to allow rav1e to access PlaneRegion memory that are outside of coded frame size

### PR DESCRIPTION
**Pull request by ycho**
The goal of this work is to establish the ground so that rav1e can encode blocks that straddle on bottom or right frame boundaries (i.e. #2396).
I thought it would better approach to first land this groundwork then land the open-partition feature,  #2396.
Because, that way we can ensure the groundwork itself does not affect the integrity of current master without any open-partition code mixed.
With this PR only, it is expected to not change coding behavior of current master.

Until it is finished, it will fail many of rav1e's internal cargol tests, mostly when comparing encoder and decoder outputs, not because the enc and dec's outputs really unmatched but more likely the dimension of plane when it is declared is not correctly changed.
When compare the two reconstructed outputs from rav1e and decoder that run on command line and not via cargo test , I see that they are matched.
Also, at the moment I create this PR, I don't observe rust panic caused by memory access violation (i.e. accessing outside the bounding box defined by PlaneRegion (+ subregion) ).

### CODE REVIEW

File: v_frame/src/frame.rs [hunk 1]
```diff
@@ -27,8 +27,9 @@ impl<T: Pixel> Frame<T> {
     width: usize, height: usize, chroma_sampling: ChromaSampling,
     luma_padding: usize,
   ) -> Self {
-    let luma_width = width.align_power_of_two(3);
-    let luma_height = height.align_power_of_two(3);
+    //TODO: If SB size is 128x128, 7 should be used instead of 6
```
Comment (vibhoothiiaanand) [inline v_frame/src/frame.rs:30]: Hi,

I was thinking whether we should keep separate commit for v_frame changes as it is a separate module and we do maintain releases for them.
If it is split as an extra commit for v_frame changes, we can make things easier,

But I am not sure if it will break the intermediate commits or not, if that is the case, we should be having this style only,

**Comment by ycho** [comment]
One tip for me: To test one failed test,

env RUST_BACKTRACE=full RAV1E_CPU_TARGET=rust cargo test --features=decode_test_dav1d dimension_32x32_dav1d

**Comment by ycho** [comment]
Hmm, after I see compare_pic( ) (used during test, comparing encoder and decoder output frames),  I start to think that the idea of rounding up input and output frame (i.e. their planes) size to SB might be bad.

**Comment by ycho** [comment]
In Irc #daala,
xiphmont, Another thing I am seeing now that is surely related to work you did before/do later is the use of bounding box in cdef_filter_tile( ) via rect( ).width|height.

<codeview> This makes me look again what I think so far is reasonable. I.e. TileRect is rounded up to SB size (thus its input and output (i.e. rec) tiled planes (i.e. PlaneRegion) ) but ts.width|height is not rounded up and keep coded w and h of tile.)  With my approach, In cdef_filter_tile( ),  "rec_w = rec.planes[p].rect().width" cause w and h include outside of coded frame w and h, which is wrong with current code of the  function.

<codeview> So, I am basically needing more time on pondering which way is more correct, tiled plane following coded w and h as current master or extend it include the block crossing frame boundaries. The most basic clue to me is that, as with the case how CFL works with block straddle on frame boundaries, if a block being encoded for any prediction mode requires to read/write pixels outside coded frame, the area should be included  as a plane.

**Comment by ycho** [comment]
YAY~~~~~~~!!!

File: v_frame/src/frame.rs (hunk 1 above)
Comment (ycho) [reply v_frame/src/frame.rs]: This change has been removed.

**Comment by ycho** [comment]
There is no change in coding efficiency or encoding speed.
[awcy, -s 6](https://beta.arewecompressedyet.com/?job=master_c89bf8bd_s6_low_latency%402020-07-21T08%3A39%3A47.175Z&job=fix_access_outside_frame_s6_low_latency%402020-07-21T08%3A41%3A00.423Z)

**Comment by ycho** [comment]
With my most recent commit(s), I have focused on correctly replace the existing bounding 'rectangle' under PlaneRegion (either from input or output tile planes) with TileState.input_tile|res.plane[ ].plane_cfg.width|height or TIleState.width|height if possible. In some cases, correct bounding rectangle requires to apply the origin of rect, i.e. rect.x|y, because the origin of rectangle can be off the (0,0), as in two cases I see, 1) via subregion(), 2) in cargo test, 'tile_encoding_with_stretched_restoration_units' .

**Comment by ycho** [comment]
As soon as this PR has landed, I will rebase my open-partition PR https://github.com/xiph/rav1e/pull/2396 on it then ask reviewers, esp. @lu_zero, to review again.

File: src/tiling/tiler.rs [hunk 2]
```diff
@@ -460,9 +460,9 @@ pub mod test {
     assert_eq!((32, 0, 32, 32), rect(&tile.planes[2]));
 
     let tile = &tile_states[2].rec; // the top-right tile
```
Comment (KyleSiefring) [inline src/tiling/tiler.rs:462]: Change comments above.

Comment (ycho) [reply src/tiling/tiler.rs:462]: Sure, which format of doc would you think best?  New *.md file under doc directory?

File: src/partition.rs [hunk 3]
```diff
@@ -670,8 +670,13 @@ pub fn get_intra_edges<T: Pixel>(
         partition_bo.0.x > 0
       };
 
-    let right_available = x + tx_size.width() < dst.rect().width;
-    let bottom_available = y + tx_size.height() < dst.rect().height;
+    let rect_w =
```
Comment (ycho) [inline src/partition.rs:673]: The rect_w, rect_h are derived from coded tile size. I.e. exclude the out of frame region because dst.rect().width|height now include out of frame region to access pixels belong to a block on frame boundary.

Comment (ycho) [reply src/partition.rs]: > The rect_w, rect_h are derived from coded tile size. I.e. exclude the out of frame region because dst.rect().width|height now include out of frame region to access pixels belong to a block on frame boundary.

I might want to add member functions under rect of PlaneRegion, which returns these two values.

File: src/tiling/tiler.rs (hunk 2 above)
Comment (KyleSiefring) [reply src/tiling/tiler.rs:462]: These are the comments I meant. https://github.com/xiph/rav1e/pull/2445/files#diff-61a9dba6df92b3be4f74e67feaa6cd10R448

I think a new md file would be great.

File: src/tiling/tile_state.rs [hunk 4]
```diff
@@ -126,24 +126,24 @@ impl IndexMut<usize> for MiTileState {
 impl<'a, T: Pixel> TileStateMut<'a, T> {
   pub fn new(
     fs: &'a mut FrameState<T>, sbo: PlaneSuperBlockOffset,
-    sb_size_log2: usize, width: usize, height: usize,
+    sb_size_log2: usize, w: usize, h: usize,
```
Comment (barrbrain) [inline src/tiling/tile_state.rs:129]: It appears that reverting the rename of the `width` and `height` parameters will result in a smaller, more focused set of changes to this function.

Comment (ycho) [reply src/tiling/tile_state.rs:129]: Yes, that is surely right. I also wanted to change as small as possible but it seems the lines 142 and 143 prohibits it by a rust grammer. I did end up compile errors. I thought one.possible way to keep formal parameter names width and height here would be change the field names width,height of TileRect struct.

Comment (vibhoothiiaanand) [reply src/tiling/tile_state.rs:129]: We could just keep as
Width.... instead of explicitly assigning right? 
I think if we are passing all args we could do like that 
If so it might work, not so sure

Comment (ycho) [reply src/tiling/tile_state.rs:129]: By introducing two new variables to keep width and height that are rounded up to SB, I think I can keep existing width and height. Let me try..

File: src/tiling/tile_state.rs [hunk 5]
```diff
@@ -136,11 +136,15 @@ impl<'a, T: Pixel> TileStateMut<'a, T> {
       height % MI_SIZE == 0,
       "Tile width must be a multiple of MI_SIZE"
     );
+
+    let sb_rounded_width = width.align_power_of_two(sb_size_log2);
```
Comment (ycho) [inline src/tiling/tile_state.rs:140]: Yeap,  how dumb I am I have not done this way from the start and just use them to change two lines below!

File: src/tiling/tiler.rs (hunk 2 above)
Comment (ycho) [reply src/tiling/tiler.rs:462]: I've pushed document for this PR, under name "doc/TILE_ON_FRAME_BOUNDARY.md"
Please let me know if you'd like more revisions or it looks fine for landing.

**Comment by ycho** [comment]
@xiphmont, @lu_zero, Your opinions are greatly appreciately as well!

**Comment by xiphmont** [comment]
If I'm reading the code correctly, this has no effect whatsoever except at the right and bottom visible frame boundaries.  Tile access still does not overlap.

So... Not the way I'm chasing it in my Rectangle Reform patches, but this is simple, tiny, and works.  I see no reason not to land it ASAP.

r+me

File: doc/TILE_ON_FRAME_BOUNDARY.md [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+## Accessing Tiled-pixels outside of a frame
+
+Tiled input and output planes (type PlaneRegion) are configured in order to
+allow memory access of its pixels that are located outside of frame boundary.
+This configuration is only effective when the tile is adjacent to right or bottom frame boundary and
+a partition block straddles on any frame boundary.
+
+The type names of tiled input and output in the codebase is defined in:
+
+"'TileStateMut' struct"
```
Comment (vibhoothiiaanand) [inline doc/TILE_ON_FRAME_BOUNDARY.md:12]: Nit: Can we use code and code blocks for this part so it looks nicer.

File: doc/TILE_ON_FRAME_BOUNDARY.md [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+"  ..."
+"  pub input: &'a Frame<T>,     // the whole frame"
+"  pub rec: TileMut<'a, T>,"
+"  ..."
+"}"
+
+## How bounding box is enforced to a PlaneRegion and now it allows to access outside of frame pixels
+
+A memory access to tiled input and output pixels in a _tile_ 
+is bounded by 'rect' field of each plane (of type 'PlaneRegion') of the 'input_tile' or 'rec' of 'TileStateMut'.
```
Comment (vibhoothiiaanand) [inline doc/TILE_ON_FRAME_BOUNDARY.md:23]:
```suggestion
is bounded by 'rect' field of each plane (of type `PlaneRegion`) of the `input_tile` or `rec` of `TileStateMut`.
```

So it will be better when we render the markdown :)

File: doc/TILE_ON_FRAME_BOUNDARY.md (hunk 6 above)
Comment (ycho) [reply doc/TILE_ON_FRAME_BOUNDARY.md:12]: Sur- Is ``` for code block?

File: doc/TILE_ON_FRAME_BOUNDARY.md (hunk 7 above)
Comment (ycho) [reply doc/TILE_ON_FRAME_BOUNDARY.md]: Done. Thanks!

File: src/tiling/tiler.rs (hunk 2 above)
Comment (lu-zero) [reply src/tiling/tiler.rs:462]: @KyleSiefring is that enough for you?
